### PR TITLE
Issue 53 timeseries primitives

### LINKED ISSUE #53: Primitive for timeseries data manipulation
Add primitives to aggregate and generate window sequences from time series data.

I will add two methods to `mlprimitives/timeseries.py`:
- create_window_sequences
        Function that takes in a Pandas.DataFrame and a window_size then creates output arrays that correspond to a timeseries sequence with window_size overlap. The output arrays can be fed into a timeseries forecasting model.
        Inputs:
            df_timeseries (Pandas.DataFrame): a Pandas dataframe which has 'timestamp' and 'value' columns, and is sorted based on timestamp. The timestamp column is in UNIX format (in seconds).
            window_size (int): number of values that overlap to create the sequence.
        Outputs:
            x (numpy.ndarray): contains the time series sequenced data.
            y (numpy.ndarray): acts as the label for the forecasting problem.
            time (numpy.ndarray): the corresponding timestamps series.

- aggregate_average_time
        Function that aggregates data in a Pandas dataframe by averaging over a given interval. It starts averaging from specified start_time.
        Inputs:
            df_time_value (Pandas.DataFrame): a Pandas dataframe which has 'timestamp' and 'value' columns, and is sorted based on timestamp. The timestamp column is in UNIX format (in seconds).
            interval_time_delta (int): an Integer denoting the number of seconds in the desired interval.
            start_time (int): a UNIX time stamp indicating the time to start aggregating. Can be smaller than the smallest time stamp value in the dataframe.
            end_time (int): a UNIX time stamp indicating the time to end aggregating. Can be larger than the largest time stamp value in the dataframe. 
        Outputs:
            aggregated_df (Pandas.DataFrame): a Pandas dataframe with two colums ('timestamp' and 'value'), where each `timestamp` is the starting time of an interval and the `value` is the result of aggregation. For intervals that don't have data in df_time_value but are still included in start_time and end_time then the value will be NaN.

### PR DESCRIPTION
**Pull request by itinawi**
Resolve #53

### CODE REVIEW

File: mlprimitives/timeseries.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+            time (numpy.ndarray): the corresponding timestamps series.
+    """
+    X = []
+    Y = []
+    time = []
+    for i in range(len(df_timeseries) - window_size):
+        X.append(df_timeseries[i: i + window_size]['value'].values.copy().reshape([-1, 1]))
+        Y.append(df_timeseries[i + 1: i + window_size + 1]['value'].values.copy().reshape([-1, 1]))
+        time.append(df_timeseries.iloc[i + window_size]['timestamp'])
+    return np.asarray(X), np.asarray(Y), np.asarray(time)
```
Comment (csala) [inline mlprimitives/timeseries.py]: Would you mind adding a line break before this line? This is just to be consistent with the styling of the rest of the library.

File: mlprimitives/timeseries.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
+import pandas as pd
+import numpy as np
+
+
+def create_window_sequences(df_timeseries, window_size):
```
[the remark this answers is not in the document]
Comment (csala) [reply mlprimitives/timeseries.py]: Please also call this something like `rolling_window` to make it more descriptive. I would also avoid the `create_` bit.

File: mlprimitives/timeseries.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    """
+        Function that takes in a Pandas.DataFrame and a window_size then creates
+            output arrays that correspond to a timeseries sequence with window_size overlap.
+            The output arrays can be fed into a timeseries forecasting model.
+        Inputs:
```
Comment (csala) [inline mlprimitives/timeseries.py]: Please follow the google docstring style: `Args` instead of `Inputs` and `Returns` instead of `Outputs`.
Also, in the `Returns` section you only specify the type, not the name.

File: mlprimitives/timeseries.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    while start_ts < end_time:
+        # average the values between start_ts, [start_ts + timedelta (e.g. 6hrs)]
+        upper_ts = start_ts + interval_time_delta
+        mask = (df_time_value['timestamp'] > start_ts) & (df_time_value['timestamp'] <= upper_ts)
+        average_value = df_time_value.loc[mask]['value'].mean(skipna=True)
+
+        accepted_points.append([start_ts, average_value])
+        start_ts = upper_ts  # update the timestamp
+
+    new_df = pd.DataFrame(accepted_points, columns=['timestamp', 'value'])
```
Comment (csala) [inline mlprimitives/timeseries.py]: You can skip the last line and directly `return pd.DataFrame...`

File: mlprimitives/timeseries.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    Y = []
+    time = []
+    for i in range(len(df_timeseries) - window_size):
+        X.append(df_timeseries[i: i + window_size]['value'].values.copy().reshape([-1, 1]))
+        Y.append(df_timeseries[i + 1: i + window_size + 1]['value'].values.copy().reshape([-1, 1]))
+        time.append(df_timeseries.iloc[i + window_size]['timestamp'])
+    return np.asarray(X), np.asarray(Y), np.asarray(time)
+
+
+def aggregate_average_time(df_time_value, interval_time_delta, start_time, end_time):
```
Comment (csala) [inline mlprimitives/timeseries.py]: I would use a more descriptive name like `average_by_time`, or `time_segments_average`.

Comment (csala) [reply mlprimitives/timeseries.py]: Same thing as above about the `df_` prefix. Also the same for the `_time_delta` suffix. Call them `time_value` and `interval`.
Or, better, rename `time_value` to `X`.

Comment (csala) [reply mlprimitives/timeseries.py]: Also add `value_column` and `time_column` as input arguments.

File: mlprimitives/timeseries.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        Function that aggregates data in a Pandas dataframe by averaging over a given interval. 
+            It starts averaging from specified start_time.
+        Inputs:
+            df_time_value (Pandas.DataFrame): a Pandas dataframe which has 'timestamp' 
+                and 'value' columns, and is sorted based on timestamp. The timestamp
+                column is in UNIX format (in seconds).
+            interval_time_delta (int): an Integer denoting the number of seconds 
+                in the desired interval.
+            start_time (int): a UNIX time stamp indicating the time to start
+                aggregating. Can be smaller than the smallest time stamp value in the dataframe.
```
Comment (csala) [inline mlprimitives/timeseries.py]: > Can be smaller than the smallest time stamp value in the dataframe.
Is this right? If I'm reading it correctly, the current implementation would generate empty rows in the dataframe if this happens.
The same happens with the end_date.

File: mlblocks_primitives/mlprimitives.timeseries.aggregate_average_time.json [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+        "type": "preprocessor",
+        "subtype": "feature_extractor"
+    },
+    "modalities": ["timeseries"],
+    "primitive": "mlprimitives.timeseries.aggregate_average_time",
+    "produce": {
+        "args": [
+            {
+                "name": "df_time_value",
+                "type": "Pandas.DataFrame"
```
Comment (csala) [inline mlblocks_primitives/mlprimitives.timeseries.aggregate_average_time.json]: call this "X" and use the "keyword" entry to make it match the actual argument name.
Also use lowercase for `pandas`.
So:
```
{
    "name": "X",
    "keyword": "time_value",
    "type": "pandas.DataFrame"
}
```

File: mlblocks_primitives/mlprimitives.timeseries.aggregate_average_time.json [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+            },
+            {
+                "name": "interval_time_delta",
```
Comment (csala) [inline mlblocks_primitives/mlprimitives.timeseries.aggregate_average_time.json]: This should be set as a fixed hyperparameter instead of a produce argument.

File: mlprimitives/timeseries.py (hunk 5 above)
Comment (csala) [reply mlprimitives/timeseries.py]: Also, is it necessary to make start_time and end_time mandatory? What if we just want to process the whole input dataframe, from start to end, without having to worry about which dates we have?
Please make them optional (default=None) and adapt the code accordingly.

File: mlblocks_primitives/mlprimitives.timeseries.aggregate_average_time.json [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+                "type": "int"
+            },
+            {
+                "name": "start_time",
```
Comment (csala) [inline mlblocks_primitives/mlprimitives.timeseries.aggregate_average_time.json]: This and end_time should be set as fixed hyperparameters with `null` as default.

File: mlblocks_primitives/mlprimitives.timeseries.create_window_sequences.json [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    "modalities": ["timeseries"],
+    "primitive": "mlprimitives.timeseries.create_window_sequences",
+    "produce": {
+        "args": [
+            {
+                "name": "df_timeseries",
+                "type": "Pandas.DataFrame"
+            },
+            {
+                "name": "window_size",
```
Comment (csala) [inline mlblocks_primitives/mlprimitives.timeseries.create_window_sequences.json]: This should be set as a fixed hyperparameter.

File: mlprimitives/timeseries.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (csala) [reply mlprimitives/timeseries.py]: Also please pass the names of the columns as arguments value_column and time_column, even if they have as default value the ones that you used.
These will also need to be set in the JSON annotation as fixed hyperparameters.

File: mlprimitives/timeseries.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+            x (numpy.ndarray): contains the time series sequenced data.
+            y (numpy.ndarray): acts as the label for the forecasting problem.
+            time (numpy.ndarray): the corresponding timestamps series.
+    """
+    X = []
+    Y = []
+    time = []
+    for i in range(len(df_timeseries) - window_size):
+        X.append(df_timeseries[i: i + window_size]['value'].values.copy().reshape([-1, 1]))
+        Y.append(df_timeseries[i + 1: i + window_size + 1]['value'].values.copy().reshape([-1, 1]))
```
Comment (csala) [inline mlprimitives/timeseries.py]: Is this implementation right? If I understand it properly, if window size is 50, each element appended to `X` will have 50 rows, which is right. But each element appended to `y` will ALSO have 50 rows, which I believe is incorrect.

File: mlprimitives/timeseries.py [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+        time.append(df_timeseries.iloc[i + window_size]['timestamp'])
```
Comment (csala) [inline mlprimitives/timeseries.py]: What do we need the `time` from this point on?

File: mlprimitives/timeseries.py (hunk 1 above)
Comment (itinawi) [reply mlprimitives/timeseries.py]: The Linter suggested I remove the line 😆

File: mlprimitives/timeseries.py (hunk 6 above)
Comment (itinawi) [reply mlprimitives/timeseries.py]: Yes, because if you want to do a multi-PID (multi-signal) LSTM then you are not sure that all the starting times would line up. Technically, if the default value for `start_time` is the smallest timestamp in the dataframe, then we can remove `start_time` and find the intervals starting from the smallest time. The same applies to `end_time` but the problem is, if we are using multi-signal LSTM then we would need to make sure the timestamps match up. What do you think?

File: mlprimitives/timeseries.py (hunk 1 above)
Comment (csala) [reply mlprimitives/timeseries.py]: That's strange. It should complain about two bank lines, but not about one.
Anyhow, let's try having one there and see it if accepts it, just to keep it consistent with the rest :-)

File: mlprimitives/timeseries.py (hunk 6 above)
Comment (csala) [reply mlprimitives/timeseries.py]: That's true. But right now this should be out of the scope of this primitive.
Let's focus on a single table and make the default values be the min and the max. If someone wants to use this in more than one signal at a time, it's their responsibility to make sure that they are aligned.

File: mlprimitives/timeseries.py (hunk 12 above)
Comment (itinawi) [reply mlprimitives/timeseries.py]: This may be used as input to other models (if we aren't using LSTM). For instance a neural network model may try to take in the time axis as a feature to learn patterns. I'm not sure if any architectures actually do this, but I want to keep this option open.
(keep in mind this isn't the timestamps of the entire dataframe but only the timestamps of the aggregated intervals).

File: mlprimitives/timeseries.py (hunk 11 above)
Comment (itinawi) [reply mlprimitives/timeseries.py]: Good point. My model was returning sequences (in case we want to stack LSTM units). But I guess this shouldn't be the default for a primitive. I will add functionality to have a default parameter such that return_sequences is set to False.

Comment (itinawi) [reply mlprimitives/timeseries.py]: Or actually I will just make it return a single Y. I don't think it's necessary to have a sequence, even if we are doing a stacked LSTM (LSTM output is input to another LSTM).

File: mlprimitives/timeseries.py [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+            time_column (string): name of column that has the time field.
+        Returns:
+            (numpy.ndarray): contains the time series sequenced data with each 
+                entry having window_size rows.
+            (numpy.ndarray): acts as the label for the forecasting problem with 
+                each entry having window_size rows.
+            (numpy.ndarray): the corresponding timestamps series.
+    """
+    output_X = []
+    Y = []
```
Comment (csala) [inline mlprimitives/timeseries.py]: Can we call it `y` in lowercase? the convention is that matrices (so, Nd, N>1) are uppercase and vectors (so, 1d) are lowercase. following the mathematical notation.

File: mlprimitives/timeseries.py [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    time = []
+    for i in range(len(X) - window_size):
+        # reshape into a vector to fit into a neural network model (vectorize it)
+        output_X.append(X[i: i + window_size][value_column].values.copy().reshape([-1, 1]))
```
Comment (csala) [inline mlprimitives/timeseries.py]: Is the `copy()` bit necessary? I think we can safely skip it and relax a bit the memory overhead.

Comment (itinawi) [reply mlprimitives/timeseries.py]: yep I think it's safe to do so, since we aren't altering the dataframe elsewhere.

**Comment by csala** [comment]
Thanks @itinawi !
